casanovo/denovo/model.py

```diff
@@ -171,16 +178,613 @@ def forward(
+        cache_tokens : torch.Tensor of shape (n_spectra * n_beams, max_length)
+            The token for each amino acid in cached peptides.
+        cache_next_idx : Dict[int, int]
+            Next available tensor index to cache peptides for each spectrum.
+        cache_pred_seq : Dict[int, Set[torch.Tensor]]
+            Set of decoded peptide tokens for each spectrum.
+        cache_pred_score : Dict[int, List[List[Tuple[float, int]]]
+            Confidence score for each decoded peptide, separated as
+            precursor m/z fitting vs not, for each spectrum.
+        """
+        batch = scores.shape[0]
+        beam = scores.shape[-1]
+
+        # Cache terminated beams and their scores
+        cache_scores = einops.rearrange(scores.clone(), "B L V S -> (B S) L V")
+        cache_tokens = einops.rearrange(tokens.clone(), "B L S -> (B S) L")
+
+        # Keep pointer to free rows in the cache and already cached predictions
+        cache_next_idx = {
+            spec_idx: spec_idx * beam for spec_idx in range(batch)
+        }
+        # Keep already decoded peptides to avoid duplicates in cache
+        cache_pred_seq = {spec_idx: set() for spec_idx in range(batch)}
+        # Store peptide scores to replace lower score peptides in cache
+        # with higher score peptides during decoding.
+        cache_pred_score = {spec_idx: [[], []] for spec_idx in range(batch)}
+
+        return (
+            cache_scores,
+            cache_tokens,
+            cache_next_idx,
+            cache_pred_seq,
+            cache_pred_score,
+        )
+
+    def _terminate_finished_beams(
+        self,
+        tokens: torch.Tensor,
+        precursors: torch.Tensor,
+        is_beam_prec_fit: List[bool],
+        idx: int,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Terminate beams exceeding precursor m/z tolerance.
+
+        Track all terminated beams.
+
+        Parameters
+        ----------
+        tokens : torch.Tensor of shape (n_spectra * n_beams, max_length)
+            Output token of the model corresponding to amino acid sequences.
+        precursors : torch.Tensor of size (n_spectra * n_beams, 3)
+            The measured precursor mass (axis 0), precursor charge (axis 1), and
+            precursor m/z (axis 2) of each MS/MS spectrum.
+        is_beam_prec_fit: List[bool]
+            List indicating if current beams are within precursor m/z tolerance.
+        idx : int
+            Index to be considered in the current decoding step.
+
+        Returns
+        -------
+        finished_beams_idx : torch.Tensor
+            Indices of all finished beams on tokens tensor.
+        tokens : torch.Tensor of size (n_spectra * n_beams, max_length)
+            Output token of the model corresponding to amino acid sequences.
+        """
+        # Check if N-terminal NH3 loss is in the vocabulary
+        is_nh3_loss_in_vocab = (
+            self.peptide_mass_calculator.masses.get("-17.027", "") != ""
```

Comment:
**todo:** This can be simplified by `"-17.027" in self.peptide_mass_calculator.masses`.